ibm_db2.stub.php

```diff
@@ -27,7 +27,7 @@ function db2_pconnect(string $database, ?string $username, ?string $password, ar
 /**
  * @param resource $connection
  */
-function db2_autocommit($connection, ?int $value = null): int|bool {}
+function db2_autocommit($connection, int $value = null): int|bool {}
```

Comment:
```suggestion
function db2_autocommit($connection, int $value): int|bool {}
```

Reply:
This parameter is optional. How do we make this explicit if we remove the default value?
https://github.com/php/pecl-database-ibm_db2/blob/07750cb56e3a873ca332702b64bfa00fbaab82ae/ibm_db2.stub.php#L5

Reply:
if this is not nullable, we can't set the default to null, can we?

Reply:
Currently, the only way in PHP to allow optional parameters is to defining default values for them. Generally, the default value used when the intention is to represent the absence of a value, is `null`.
See https://3v4l.org/2JIXi.

Reply:
I get that, but as you can see (https://3v4l.org/utCc3), if you don't specify a value, the default will be NULL, which in turn would result in an error when called this way.

Reply:
I'm sorry, I think I'm reading this in the wrong way or I don't get your point.
The stub in this PR is declaring `null` as default value not because the default value in the function is NULL, but because this is the sanest way we have to mark a parameter as optional.
The C implementation first check the number of received arguments before validating the value received in the argument 2:
https://github.com/php/pecl-database-ibm_db2/blob/75b16f651e671ab07d581b063302440aaa5939ad/ibm_db2.c#L2922-L2923

Reply:
I was just saying that if you were to write a wrapper function for db2_autocommit like that (with `$connection, int $value = null` as the arguments), the call to db2_autocommit would fail without explicitly setting the 2nd argument every single time.

If the stub is just used for creating documentation, then it might be ok, but in that case I'd rather use something this (which makes semantically more sense):

```
function db2_autocommit($connection, int $value = DB2_AUTOCOMMIT_ON|DB2_AUTOCOMMIT_OFF): int|bool {}
```

Reply:
AFAIK, this stub is only used for documenting the function signature, but the stub must be written using valid PHP 8 syntax:
https://github.com/php/pecl-database-ibm_db2/blob/07750cb56e3a873ca332702b64bfa00fbaab82ae/ibm_db2.stub.php#L4

Regarding your suggestion, I'm pretty sure the current GA versions of PHP do not support union values (`DB2_AUTOCOMMIT_ON|DB2_AUTOCOMMIT_OFF`) in function arguments.

Reply:
I think the stub could be written with a newer version; I think the generator should make a header compatible with all 8.x versions (and 7.x in the other file). I remember generating it on 8.1 when I did make changes.

If we want to make `autocommit(conn, NULL)` work, the function could also be changed to check for a null zval instead of just argc.

Reply:
Yep, I think that sending NULL as second argument should be handled the same as if it was not sent in the first place. I can make the changes if that's ok.

Reply:
@NattyNarwhal let me know, if I can make the change to the `db2_autocommit` function. The change is trivial, but I will have to setup a new dev VM, since I will also have to adjust and add new tests.

Reply:
>I think the stub could be written with a newer version; I think the generator should make a header compatible with all 8.x versions (and 7.x in the other file). I remember generating it on 8.1 when I did make changes.

Do you mean regenerating the contents in `ibm_db2_arginfo.h` from the latest version of `ibm_db2.stub.php` in the `master` branch or from a specific release? (like the work done in php/doc-en#2316).
Should we add a stub file (`ibm_db2_legacy.stub.php` or similar) for `ibm_db2_legacy_arginfo.h` using PHP 7.x?

Please, let me know if I am understanding correctly, because I never used the `gen_stub.php` script before.

>If we want to make autocommit(conn, NULL) work, the function could also be changed to check for a null zval instead of just argc.

Regarding this, I think I like the strictness the current implementation has (allowing only the meaningful values), but since this can not be represented from a PHP function signature, I have no personal preference about this decision.

Reply:
@tessus  Go for it if you want.

I think it's worth asking upstream what the backwards compatibility is for generated arginfo headers for a stub. I know the legacy header it generates is always going to be 7.4 compatible, but are there any guarantees for, say, if I generate on 8.3, it'll build (and work) on 8.0?

Reply:
> Regarding this, I think I like the strictness the current implementation has (allowing only the meaningful values), but since this can not be represented from a PHP function signature, I have no personal preference about this decision.

The strictness wouldn't change. The change would just make `db2_autocommit(conn)` the same as `db2_autocommit(conn, NULL)` without throwing an error. If anything other than NULL was passed as an argument, it would still require either `DB2_AUTOCOMMIT_ON` or `DB2_AUTOCOMMIT_OFF`.

Reply:
@NattyNarwhal ok, I'll setup the VM first. ;-) since this doesn't have to be added to this PR, I'll open a separate one. I should be able to get to it within the next few days.

Reply:
>The strictness wouldn't change. The change would just make db2_autocommit(conn) the same as db2_autocommit(conn, NULL) without throwing an error. If anything other than NULL was passed as an argument, it would still require either DB2_AUTOCOMMIT_ON or DB2_AUTOCOMMIT_OFF.

I mean, allowing `null` in the argument 2 will add a new supported type on it.

Reply:
Ok, let me go through the stub again and check the other functions. I just wanted to make the function semantically correct for what is shown in the stub, because right now it is not.
According to the stub, one can send NULL, because NULL would be sent by default if no 2nd argument was passed. And this is just not true.

Maybe I have to take a step back and re-evaluate my understanding of the stub.

Reply:
I'll also note that my translation from the old way of declaring functions to the stub might not be perfect. I tested against the test suite and user reports, but there's definitely the possibility I got some of the signatures wrong.

Reply:
>I'll also note that my translation from the old way of declaring functions to the stub might not be perfect. I tested against the test suite and user reports, but there's definitely the possibility I got some of the signatures wrong.

As far as I can see, there are no problems about the way that the signatures are represented regarding the PHP syntax.

One thing to mention: Even if the language interpreter allows this behavior near nullable arguments without notice, I think we SHOULD use the available syntax following this convention:

1. Argument 1 MUST be passed, allowing `int` or `null`:
```php
function test(?int $param) {};
```

2. Argument 1 is optional, and allows `int` or `null`:
```php
function test(?int $param = null) {};
```

3. Argument 1 is optional, and allows only `int` (the case in `db2_autocommit()`):
```php
function test(int $param = null) {};
```

Reply:
If an optional parameter doesn't have a valid default value then you can use the `UNKNOWN` pseudo default value.  php-src has quite a few of such parameters (although most of them will be phased out if my RFC in progress will have accepted).

Reply:
Thank you so much for the help @kocsismate!
 I've updated the PR with your suggestion.